state-sync/storage-service/server/src/lib.rs

```diff
@@ -0,0 +1,238 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+use diem_infallible::RwLock;
+use diem_types::{epoch_change::EpochChangeProof, transaction::TransactionListWithProof};
+use serde::{Deserialize, Serialize};
+use std::sync::Arc;
+use storage_interface::DbReaderWriter;
+use storage_service_types::{
+    DataSummary, EpochEndingLedgerInfoRequest, ProtocolMetadata, ServerProtocolVersion,
+    StorageServerSummary, StorageServiceRequest, StorageServiceResponse,
+    TransactionsWithProofRequest,
+};
+use thiserror::Error;
+
+#[cfg(test)]
+mod tests;
+
+// TODO(joshlind): make these configurable.
+/// Storage server constants.
+pub const MAX_TRANSACTION_CHUNK_SIZE: u64 = 1000;
+pub const STORAGE_SERVER_VERSION: u64 = 1;
+
+#[derive(Clone, Debug, Deserialize, Error, PartialEq, Serialize)]
+pub enum Error {
+    #[error("Storage error encountered: {0}")]
+    StorageErrorEncountered(String),
+    #[error("Unexpected error encountered: {0}")]
+    UnexpectedErrorEncountered(String),
+}
+
+/// The server-side implementation of the storage service. This provides all the
+/// functionality required to handle storage service requests (i.e., from clients).
+pub struct StorageServiceServer<T> {
+    storage: T,
+}
+
+impl<T: StorageReaderInterface> StorageServiceServer<T> {
+    pub fn new(storage: T) -> Self {
+        Self { storage }
+    }
+
+    pub fn handle_request(
+        &self,
+        request: StorageServiceRequest,
+    ) -> Result<StorageServiceResponse, Error> {
```

Comment:
I guess we probably want `handle_request` to return just `StorageServiceResponse` and convert any errors into `StorageServiceError`s? Even if DiemDB returns an error, it should probably just return an "Internal Error" variant

Reply:
I've done this for now but we'll need to give it more attention in the future. For example, if some errors are really bad (e.g., internal invariant violations), we may want them to bubble up to main to cause the node to die. But we can deal with that later 😄